File: packages/python/onnx9000-core/src/onnx9000/core/registry.py
"""Operator Registry.

Central catalog for registering ONNX operators across the ecosystem.
"""

from typing import Any, Callable, Optional

from onnx9000.core.exceptions import UnsupportedOpError


class OperatorRegistry:
    """Registry for mapping ONNX operator types to their respective.

    implementations, parsers, or execution kernels.
    """

    def __init__(self) -> None:
        """Initialize the operator registry."""
        self._registry: dict[str, Any] = {}
        self._domains: list[str] = [
            "",
            "ai.onnx",
            "ai.onnx.ml",
            "ai.onnx.preview.training",
            "ai.onnx.contrib",
        ]

    def register_op(
        self, domain: str, op_type: str, provider: Optional[str] = None
    ) -> Callable[[Any], Any]:
        """Decorate to register a class or function for a specific ONNX op_type."""

        def wrapper(impl: Any) -> Any:
            """Wrap the implementation and register it."""
            if domain not in self._domains:
                self._domains.append(domain)
            key = (domain, op_type, provider)
            self._registry[key] = impl
            return impl

        return wrapper

    def get_op(self, domain: str, op_type: str, provider: Optional[str] = None) -> Any:
        """Retrieve the registered implementation for an ONNX operator."""
        key = (domain, op_type, provider)
        if key not in self._registry:
            # Fallback to no provider if specific one not found
            fallback_key = (domain, op_type, None)
            if fallback_key in self._registry:
                return self._registry[fallback_key]
            raise UnsupportedOpError(f"{domain}.{op_type} (provider={provider})")
        return self._registry[key]

    def get_all_registered(self, provider: Optional[str] = None) -> dict[str, Any]:
        """Return all registered operators for a given provider."""
        return {
            f"{k[0]}.{k[1]}" if k[0] else k[1]: v
            for k, v in self._registry.items()
            if k[2] == provider
        }


global_registry = OperatorRegistry()


def register_op(domain: str, op_type: str, provider: Optional[str] = None) -> Callable[[Any], Any]:
    """Exposed decorator for registering ops."""
    return global_registry.register_op(domain, op_type, provider=provider)

File: packages/js/core/src/ops/registry.ts
/* eslint-disable */
/**
 * Operator Registry for JS Core.
 */

import { Tensor } from '../ir/tensor.js';

/**
 * Supported attribute value types in ONNX.
 */
export type AttributeValue =
  | number
  | string
  | boolean
  | number[]
  | string[]
  | boolean[]
  | Tensor
  | Tensor[]
  | undefined;

/**
 * Interface for operator implementations.
 */
export interface OpImplementation {
  /**
   * Execute the operator.
   * @param inputs Array of input tensors.
   * @param attributes Dictionary of operator attributes.
   * @returns Array of output tensors.
   */
  execute(inputs: Tensor[], attributes: Record<string, AttributeValue>): Tensor[];
}

/**
 * Registry for mapping ONNX operator types to their implementations.
 */
class OperatorRegistry {
  private registry = new Map<string, new () => OpImplementation>();

  /**
   * Register an operator.
   * @param domain Operator domain.
   * @param opType Operator type name.
   * @param provider Hardware provider name.
   */
  register_op(domain: string, opType: string, provider: string | null = null) {
    const key = this.getKey(domain, opType, provider);
    return (target: new () => OpImplementation) => {
      this.registry.set(key, target);
    };
  }

  /**
   * Get an operator implementation.
   * @param domain Operator domain.
   * @param opType Operator type name.
   * @param provider Hardware provider name.
   */
  get_op(
    domain: string,
    opType: string,
    provider: string | null = null,
  ): (new () => OpImplementation) | undefined {
    const key = this.getKey(domain, opType, provider);
    let impl = this.registry.get(key);
    if (!impl && provider !== null) {
      // Fallback to no provider
      impl = this.registry.get(this.getKey(domain, opType, null));
    }
    return impl;
  }

  private getKey(domain: string, opType: string, provider: string | null): string {
    return `${domain || 'ai.onnx'}::${opType}${provider ? `::${provider}` : ''}`;
  }

  /**
   * Get all registered operators for a provider.
   * @param provider Hardware provider name.
   */
  getAllRegistered(provider: string | null = null): Record<string, new () => OpImplementation> {
    const result: Record<string, new () => OpImplementation> = {};
    for (const [key, value] of this.registry.entries()) {
      const parts = key.split('::');
      const p = parts.length > 2 ? parts[2] : null;
      if (p === (provider || null)) {
        result[`${parts[0]}::${parts[1]}`] = value;
      }
    }
    return result;
  }
}

/**
 * Global operator registry instance.
 */
export const globalRegistry = new OperatorRegistry();

/**
 * Decorator to register an operator.
 * @param domain Operator domain.
 * @param opType Operator type name.
 * @param provider Hardware provider name.
 */
export function register_op(domain: string, opType: string, provider: string | null = null) {
  return globalRegistry.register_op(domain, opType, provider);
}
